python/src/iceberg/avro/decoder.py

```diff
@@ -0,0 +1,193 @@
+        try:
+            decimal.getcontext().prec = precision
+            scaled_datum = decimal.Decimal(unscaled_datum).scaleb(-scale)
+        finally:
+            decimal.getcontext().prec = original_prec
+        return scaled_datum
+
+    def read_bytes(self) -> bytes:
+        """
+        Bytes are encoded as a long followed by that many bytes of data.
+        """
+        return self.read(self.read_long())
+
+    def read_utf8(self) -> str:
+        """
+        A string is encoded as a long followed by
+        that many bytes of UTF-8 encoded character data.
+        """
+        return self.read_bytes().decode("utf-8")
+
+    def read_date_from_int(self) -> datetime.date:
+        """
+        int is decoded as python date object.
+        int stores the number of days from
+        the unix epoch, 1 January 1970 (ISO calendar).
+        """
+        days_since_epoch = self.read_int()
+        return datetime.date(1970, 1, 1) + datetime.timedelta(days_since_epoch)
+
+    def _build_time_object(self, value: int, scale_to_micro: int) -> datetime.time:
+        value = value * scale_to_micro
+        value, microseconds = divmod(value, 1000000)
+        value, seconds = divmod(value, 60)
+        value, minutes = divmod(value, 60)
+        hours = value
+
+        return datetime.time(hour=hours, minute=minutes, second=seconds, microsecond=microseconds)
+
+    def read_time_millis_from_int(self) -> datetime.time:
+        """
+        int is decoded as python time object which represents
+        the number of milliseconds after midnight, 00:00:00.000.
+        """
+        milliseconds = self.read_int()
+        return self._build_time_object(milliseconds, 1000)
+
+    def read_time_micros_from_long(self) -> datetime.time:
+        """
+        long is decoded as python time object which represents
+        the number of microseconds after midnight, 00:00:00.000000.
+        """
+        microseconds = self.read_long()
+        return self._build_time_object(microseconds, 1)
+
+    def read_timestamp_millis_from_long(self) -> datetime.datetime:
+        """
+        long is decoded as python datetime object which represents
+        the number of milliseconds from the unix epoch, 1 January 1970.
+        """
+        timestamp_millis = self.read_long()
+        timedelta = datetime.timedelta(microseconds=timestamp_millis * 1000)
+        unix_epoch_datetime = datetime.datetime(1970, 1, 1, 0, 0, 0, 0, tzinfo=datetime.timezone.utc)
+        return unix_epoch_datetime + timedelta
+
+    def read_timestamp_micros_from_long(self) -> datetime.datetime:
+        """
+        long is decoded as python datetime object which represents
+        the number of microseconds from the unix epoch, 1 January 1970.
+        """
+        timestamp_micros = self.read_long()
+        timedelta = datetime.timedelta(microseconds=timestamp_micros)
+        unix_epoch_datetime = datetime.datetime(1970, 1, 1, 0, 0, 0, 0, tzinfo=datetime.timezone.utc)
```

Comment:
This value is incorrect for Iceberg because Iceberg's `timestamp` type should not have `tzinfo` set. Can you split this into `read_timestamp_micros_from_long` and `read_timestamptz_micros_from_long`?

Reply:
Added!